crates/accelerate/src/convert_2q_block_matrix.rs

```diff
@@ -32,32 +38,63 @@ pub fn blocks_to_matrix(
     py: Python,
     op_list: Vec<(PyReadonlyArray2<Complex64>, SmallVec<[u8; 2]>)>,
 ) -> PyResult<Py<PyArray2<Complex64>>> {
-    let identity = aview2(&ONE_QUBIT_IDENTITY);
-    let input_matrix = op_list[0].0.as_array();
-    let mut matrix: Array2<Complex64> = match op_list[0].1.as_slice() {
-        [0] => kron(&identity, &input_matrix),
-        [1] => kron(&input_matrix, &identity),
-        [0, 1] => input_matrix.to_owned(),
-        [1, 0] => change_basis(input_matrix),
-        [] => Array2::eye(4),
+    let identity = aview2(&ONE_QUBIT_IDENTITY).into_faer_complex();
+    let input_matrix = op_list[0].0.as_array().into_faer_complex();
+    let mut matrix: Mat<c64> = Mat::<c64>::zeros(4, 4);
+
+    match op_list[0].1.as_slice() {
+        [0] => kron(matrix.as_mut(), identity.as_ref(), input_matrix.as_ref()),
+        [1] => kron(matrix.as_mut(), input_matrix.as_ref(), identity.as_ref()),
+        [0, 1] => matrix = input_matrix.to_owned(),
+        [1, 0] => matrix = change_basis_faer(input_matrix.to_owned()),
+        [] => matrix = Mat::<c64>::identity(4, 4),
         _ => unreachable!(),
     };
     for (op_matrix, q_list) in op_list.into_iter().skip(1) {
-        let op_matrix = op_matrix.as_array();
+        let op_matrix = op_matrix.as_array().into_faer_complex();
+        let mut result: Mat<c64> = Mat::<c64>::zeros(4, 4);
 
-        let result = match q_list.as_slice() {
-            [0] => Some(kron(&identity, &op_matrix)),
-            [1] => Some(kron(&op_matrix, &identity)),
-            [1, 0] => Some(change_basis(op_matrix)),
-            [] => Some(Array2::eye(4)),
-            _ => None,
-        };
-        matrix = match result {
-            Some(result) => result.dot(&matrix),
-            None => op_matrix.dot(&matrix),
+        match q_list.as_slice() {
+            [0] => kron(result.as_mut(), identity.as_ref(), op_matrix.as_ref()),
+            [1] => kron(result.as_mut(), op_matrix.as_ref(), identity.as_ref()),
+            [1, 0] => result = change_basis_faer(op_matrix.to_owned()),
+            [] => result = Mat::<c64>::identity(4, 4),
+            _ => result = op_matrix.to_owned(),
         };
+
+        let aux = matrix.clone();
+        matmul(
+            matrix.as_mut(),
+            result.as_ref(),
+            aux.as_ref(),
+            None,
+            c64::new(1., 0.),
+            Parallelism::None,
+        );
```

Comment:
Looking at the faer docs, it looks like you may be able to just do:

```rust
let res = result * matrix;
```

https://docs.rs/faer/latest/faer/#basic-usage

It obviously doesn't give us the same level as control as calling `faer::modules::core::mul::matmul;` but I wonder if there is a performance difference with it, as I think it would be a little cleaner looking than calling matmul like this.

Reply:
Agreed, I think it looks cleaner too.

> I wonder if there is a performance difference with it

To test it, I ran the same tests as in the issue, and it seems like we lose performance if we don't call `faer::modules::core::mul::matmul;`

|       |**% change avg (asterisk -> matmul)** | asterisk (min) | asterisk (max) | **asterisk (avg)** | matmul (min) | matmul (max) | **matmul (avg)** |
|:-----:|:---------------:|:---------------:|:-------------:|:-------------------:|:------------:|:------------:|:----------------:|
|  2x2  |    **92.80%**        |            524ns     |     651ns     |    **547.67ns**   |    38ns   |    70ns   |  **39.41ns**  |
|  4x4  |    **78.19%**       |        659ns     |     1.00µs     |    **689.28ns**   |   145ns  |   217ns   |  **150.34ns**  |

Reply:
Wow, that's quite the difference, I wasn't expecting that. Then yeah lets leave it as is.

Reply:
Oh, wait what exactly were you timing here? Was it just the time to call `matmul` or did you include the `clone()` time? The bare `matmul` call by itself is definitely gonna be faster, but you still need to allocate the result space which you are in effect doing with the `clone()` call above. The multiplication operator just does that implicitly so it's not a manual operation.

Reply:
Oh, yeah, I was only timing the `matmul` call in general, without using the same matrix as `dst` either. I rewrote the test including the `clone()`. 

So for the `matmul` call, I'm timing:
```rust
let aux = dst.clone();
matmul(dst.as_mut(), a.as_ref(), aux.as_ref(), Some(ALPHA), BETA, Parallelism::None);
```

For the other case:
```rust
dst = &dst + &a * &dst;
```

`ALPHA` and `BETA` are 1 for the first case.

|       |**% change avg (asterisk -> matmul)** | asterisk (min) | asterisk (max) | **asterisk (avg)** | matmul (min) | matmul (max) | **matmul (avg)** |
|:-----:|:---------------:|:---------------:|:-------------:|:-------------------:|:------------:|:------------:|:----------------:|
|  2x2  |    **56.61%**        |            589 ns     |     868ns     |    **632.74ns**   |    263ns   |    432ns   |  **274.56ns**  |
|  4x4  |    **44.37%**       |        717ns     |     938ns     |    **763.04ns**   |   397ns  |   558ns   |  **424.51ns**  |

Reply:
Oh, yeah, I was only timing the `matmul` call in general, without using the same matrix as the destination either. I rewrote the test including the `clone()`. 

So for the `matmul` call, I'm timing:
```rust
let aux = dst.clone();
matmul(dst.as_mut(), a.as_ref(), aux.as_ref(), Some(ALPHA), BETA, Parallelism::None);
```

For the other case:
```rust
dst = &dst + &a * &dst;
```

`ALPHA` and `BETA` are 1 for the first case.

|       |**% change avg (asterisk -> matmul)** | asterisk (min) | asterisk (max) | **asterisk (avg)** | matmul (min) | matmul (max) | **matmul (avg)** |
|:-----:|:---------------:|:---------------:|:-------------:|:-------------------:|:------------:|:------------:|:----------------:|
|  2x2  |    **56.61%**        |            589 ns     |     868ns     |    **632.74ns**   |    263ns   |    432ns   |  **274.56ns**  |
|  4x4  |    **44.37%**       |        717ns     |     938ns     |    **763.04ns**   |   397ns  |   558ns   |  **424.51ns**  |